### ISSUE TITLE
Write way more tests

### ISSUE THREAD

**Issue by micahflee** [micahflee/onionshare#404]
I've done a terrible job at writing tests for OnionShare. There should be way more tests, both in the `onionshare` module (especially the `onionshare.Onion` class and the `onionshare.web` module) and in `onionshare_gui`.

I haven't written Qt tests before, but it looks like these might be helpful resources:

- http://doc.qt.io/qt-5/qtest-tutorial.html
- https://github.com/pytest-dev/pytest-qt
- http://johnnado.com/pyqt-qtest-example/

**Comment by micahflee** [comment]
Related to #164. FYI, I'm not married to nosetests, I'm happy with using whatever framework works best.

**Comment by mig5** [comment]
Hi @micahflee,

I've made limited progress with tests. I have to confess I'm too much of a noob to know what I'm doing, so I thought I'd leave something that hopefully gets you or someone better than me started.

1) Download this gist and put it in the root of the onionshare dir (this is how much of a noob I am - I couldn't figure out how to import the relative modules properly by putting it in the test dir :) )

https://gist.github.com/mig5/90986df2d5ec10d60fc9514b5d5ae501

2) in onionshare/common.py I had to hack something up for the get_resource_path() prefix in order to load locales. This was done on a Linux machine, and the if/else conditional was just not matching

```
 prefix = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))), 'share')
```

probably the test could be modified to pass the onionshare_dev_mode attribute to workaround this...

3) Finally, you can run 'python3 oniontest.py', which tests:

a) the gui loaded (which also will imply the tor connection worked, as it's part of the setUpClass that runs before any tests)

b) the window title is 'OnionShare'

c) the download container is hidden

d) the settings dialog button is enabled

e) the status bar is visible

f) the number of files in the file dialog (our test.txt) is 1

I could not for the life of me work out how to get the QPushButtons to be pressed via QtTest. That will be the real magic so that we can then tests ports are open/closed etc on share start/end events

Hopefully it inspires someone to take this further/do it properly for the UI tests.

(Side note - I get a lot of 'Resourcewarning' on unclosed files, generally locale and torrc files.. seems mostly harmless)

I hope that this isn't completely wrong and a waste of your time, but I was pleased at least that I could get some passing tests.

**Comment by mig5** [comment]
Updated the gist - I was able to get the test to click on the 'Start Sharing' button - (can see in the test output that the ephemeral onion service is in the process of starting, though I would love to assert on that stdout/capture the port and do an lsof or similar) - and asserts that the server status is 1 (working).

I couldn't find a way to make it 'wait' til the service is started (that the server_status is finished and asserting 2 STATUS_STARTED). 

baby steps.. hope it helps someone progress further.

**Comment by mig5** [comment]
I discovered qWait, which is non-blocking unlike time.sleep. Works a lot better, as it allows for proper testing of the emitted signals after pressing the Start button, verifying that it transitions to RUNNING instead of WORKING etc. The UI doesn't visually 'stall' while the test carries out.

The GUI test in #411 (you can forget the now defunct Github gist) is now actually really good: it's a full end to end run of starting and stopping a share, verifying the onion address, slug and a bunch of other stuff.

The big deal is that I also finally figured out how to obtain the port number, so now we check that after stopping the service via the UI, the socket is closed, as it should be. This actually fails right now, which is good, because it's the bug in #401! When you apply the fix from #401, the test succeeds as expected - we cannot connect. So hopefully that one can be considered unblocked.

I also added a separate onionshare_settings_test.py to #411, which tests that unchecking the 'stay open' variable correctly writes false to the json. Be careful running this on a 'real' machine as it deliberately clobbers the user's onionshare.json on each test.  It also tests that the onion is rebooted after saving the settings, reconnecting to tor successfully.

The next question - how to get Travis to run these, since I think it expects an X environment to load the GUI!

**Comment by mig5** [comment]
Updated the travisCI yml in #411 to run these tests via xvfb (along with your original nosetests). Getting bizarre errors on unable to find Python modules.. probably another path issue... Works On My Machine! (a Debian Jessie AppVM). Giving up for now..

**Comment by mig5** [comment]
I've modified travis to install newer Tor/stem for ephemeral services. It is now running the GUI test via nosetests3 and failing as expected!

https://api.travis-ci.org/jobs/235506516/log.txt?deansi=true

Not sure why your existing strings test fails now though, it succeeds when run manually.. must be getting some bleed from the other test somehow. However, at least we have Travis doing GUI tests!

There's a lot of awful QPainter and related noise - a symptom of running under xvfb and no apparent way to suppress it. They might be legit warnings about the app, but it might be due to stuff in Qt that you can't do anything about.. the important thing (to me) is that the desired test actually runs.

I'm working now on another test that actually downloads the share, so we can test that the server automatically stops as desired.

**Comment by mig5** [comment]
Huh, so that's interesting, it seems indeed that your strings test gets some stdout bleed from my earlier tests, when they are run in conjunction with nosetests or pytest:

```
_________________________________________________________________________________________ test_starts_with_empty_strings __________________________________________________________________________________________

    def test_starts_with_empty_strings():
        """creates an empty strings dict by default"""
>       assert strings.strings == {}
E       assert {'closing_aut... server', ...} == {}
E         Detailed information truncated, use "-vv" to show

test/onionshare_strings_test.py:32: AssertionError
====================================================================================== 4 failed, 13 passed in 185.11 seconds ======================================================================================
```

That's a bit annoying... I'm looking to see if there's a way to order the tests so that this doesn't happen, though it doesn't seem ideal..

**Comment by mig5** [comment]
The final version of my tests are at https://github.com/mig5/onionshare/commit/5e4769362fcbdd6c4466f7e9f8c8655b09f132f4 if anyone decides to improve on them.

There's a wrapper script you can run from the root of the dir:

```
./dev_scripts/tests # Runs the full GUI experience
```

```
./dev_scripts/tests --headless # Runs with xvfb, designed for something like TravisCI
```

The full test suite takes about 3m20s to complete on my machine, currently all passing (against master branch at time of writing)

**Comment by delirious-lettuce** [comment]
@micahflee ,

I am starting to write some non-gui tests (looks like @mig5 has started on the gui tests) to improve the overall coverage and I was wondering about using `pytest`. As you probably know, it can be used to run `unittest & nose`-style tests. But there are some useful features like [`@pytest.mark.parametrize`](https://docs.pytest.org/en/latest/parametrize.html) that are `pytest` only.

Any thoughts? Should I hold off on the `pytest` specific tests and just write more generic tests for now?

**Comment by mig5** [comment]
Although I probably won't do much more on tests, I took a look at pytest and changed my commit above's wrapper script to assume use of it. It came up in #164 and I agree it looks more powerful than nosetests (things like skip and xfail for example [1]) so +1 from me to move to it.

[1] https://docs.pytest.org/en/latest/skipping.html

**Comment by delirious-lettuce** [comment]
I've written tests for most of the functions in 'common.py', I will send a
new PR in the morning. I will try to write many more tests if @micahflee
approves of the style and 'pytest' in general.

**Comment by micahflee** [comment]
First, thanks so much @mig5 and @delirious-lettuce for writing tests! This will improve OnionShare so much. I've got a few thoughts.

@mig5, regarding your #411 PR. I still haven't had a lot of time to look at it, but at a glance it looks like a great foundation for starting to write GUI tests. Lately I've been doing much of my OnionShare development in a fedora-25 AppVM, and I had to make a few tweaks to get the tests to run (like executing `py.test-3` instead of `/usr/lib/python3/dist-packages/pytest.py`). I've made a few commits in my own https://github.com/micahflee/onionshare/tree/404_gui_tests branch, which is a fork of yours. I think there's definitely useful work in here, and when I get time I want to try to whip this into shape so that all of these tests are passing, and base the OnionShare tests off of this.

@delirious-lettuce I look forward to your PR. I definitely approve of switching to pytest. Hopefully it won't be too much work to get the GUI tests that @mig5 already wrote working with the test that you've written together in a single branch.

Now that I think about it, here are some broad goals I'd like to achieve for the tests:

* There should be a single test framework for CLI and GUI -- it looks like this won't be a problem, and all the tests should work with pytest.
* The same test suite should pass in Linux, Windows, and macOS.
* Travis-CI should be able to run, and pass, the tests -- I think getting an X environment running in Travis is just a matter of installing the right packages I think.

Regarding testing the settings, I think it shouldn't be too much work to make it so test settings json files get saved to a different location, to avoid disruption any existing settings file.

Also, I think we should consider the implications of connecting to Tor and starting onion services as part of the test suite. We obviously we need to figure out how to test this stuff somehow, here here are some thoughts.

* It might not always be reliable. Sometimes connecting to Tor might fail, or a circuit could die in the middle of a test, etc., which could result in tests occasionally failing when really they should pass.
* Creating onion services involves doing work on something like 7 different Tor nodes around the world. With Travis, this will happen automatically with every commit. Is this a waste of resources?
* Connecting to Tor and creating onion services is *slow*.

For these reasons, it seems prudent to try to reduce the amount of network usage in the tests. We could stub out connecting to Tor at all unless it's testing something specifically related to connecting to Tor. We could use `--local-only` for a lot of the tests. I wonder if it makes sense to disable-by-default tests that use the network, and only run those manually sometimes? I'm not sure, but I think it's worth discussing.

**Comment by delirious-lettuce** [comment]
@micahflee ,

Great points. You have helped clarify many questions that I was curious about!

I'll send a PR of my initial tests in the next hour or two (work-in-progress tests for `common.py`). I'm getting close to having every function in `onionshare.common.py` tested but I'm always open for input. After I get that one done I will start working on tests for some of the other files.

I've never tried to test tor before so I'm not sure which way would be best there. You are probably correct that removing the network element for some tests would be much more reliable. I'm somewhat new to tor in general so I will defer to you on those matters.

I haven't actually looked at the gui tests that @mig5 wrote yet. I'm new to testing QT gui's as well but I am interested in learning and contributing.

Thanks for your time!

**Comment by delirious-lettuce** [comment]
@micahflee ,

Also, [`Hypothesis`](https://hypothesis.readthedocs.io/en/latest/index.html) can be used with `pytest`. I've only played around with it on a couple smaller projects but it might be something to think about later (after everything/almost everything is tested).

Just wanted to leave that here in the tests discussion for future reference.

**Comment by mig5** [comment]
@micahflee my tests at time of my last commit to my fork were passing - including with --headless arg to my test script which would then run it through xvfb (as if it were in Travis). For Travis you probably just need to adjust .travisci.yml like so:

https://gist.github.com/mig5/0c4f05e41ae06631cccd030c818745d5

The only issue will be whether the Trusty environment can run the dev_scripts/tests with your path to the pytest binary as you set it. I found on Debian there was no convenient /usr/bin/pytest but I may have been looking for the wrong name.

**Comment by mig5** [comment]
Begun running pytests in Travis in #416 that skip the 'full tor' GUI tests if we're in a Travis env

**Comment by mig5** [comment]
Tails devs opened their own ticket re: writing tests for Onionshare, it might be worth keeping an eye on that at least to see if de-duplication of effort is feasible https://labs.riseup.net/code/issues/12627

**Comment by u451f** [comment]
@mig5 no the Tails tests are different. While the nosetests are code tests, the Tails tests actually test the graphical interface. Any nosetest you add to OnionShare's test/ folder, will be executed automatically when I build the Debian package, so they should already work when the package is part of Tails.

**Comment by mig5** [comment]
@u451f my point was that this ticket is tracking GUI tests (in addition to the nosetests which were already in master branch), some of which are written already in PRs referenced above, so thought you might be able to save some time and use some of what was done here :)

**Comment by mig5** [comment]
@micahflee think we can close this? We have written way more tests, and we already have a ticket for refactoring our newer tests, so can probably concentrate on that one.

**Comment by micahflee** [comment]
@mig5 yes, I declare this a success
